cosmoz-page-location.html

```diff
@@ -1,238 +1,173 @@
 <link rel="import" href="../polymer/polymer.html">
 <link rel="import" href="../iron-location/iron-location.html">
 <link rel="import" href="../iron-location/iron-query-params.html">
+<!--
+The `cosmoz-page-location` element manages binding to and from the current URL.
 
+@demo demo/location.html
+-->
 <dom-module id="cosmoz-page-location">
 	<template>
-		<iron-location url-space-regex="^$"
-			path="{{appPath}}"
-			query="{{_urlQueryString}}"
-			hash="{{_urlHash}}">
+		<iron-location url-space-regex="^$" id="location"
+			path="{{ appPath }}"
+			query="{{ _appQueryString }}"
+			hash="{{ _appHashString }}">
 		</iron-location>
-		<iron-query-params
-			params-string="{{_urlQueryString}}"
-			params-object="{{appQueryParams}}">
-		</iron-query-params>
-		<iron-query-params
-			params-string="{{routeQueryString}}"
-			params-object="{{routeQueryParams}}">
-		</iron-query-params>
-		<iron-query-params
-			params-string="{{routeHash}}"
-			params-object="{{routeHashParams}}">
-		</iron-query-params>
-
-		<iron-query-params
-			params-string="{{_tempParamsString}}"
-			params-object="{{_tempParamsObject}}"></iron-query-params>
+
+		<iron-query-params id="params"
+		></iron-query-params>
+
 	</template>
 	<script>
 		(function () {
 			'use strict';
 
+			const defaultObject = function () {
+					return {};
+				},
+				assign = Object.assign;
+
 			Polymer({
 				is: 'cosmoz-page-location',
 				properties: {
-					/**
-					 *
-					 */
 					appPath: {
 						type: String,
 						notify: true
 					},
 
-					/**
-					 *
-					 */
-					appQueryParams: {
-						type: Object,
-						notify: true
+					_appQueryString: {
+						type: String
 					},
 
-					/**
-					 */
-					routePath: {
-						type: String,
-						notify: true
+					_appHashString: {
+						type: String
 					},
 
-					/**
-					 */
-					routeQueryString: {
-						type: String
+					routePath: {
+						type: String,
+						notify: true,
+						observer: '_routeChanged'
 					},
 
-					/**
-					 *
-					 */
-					routeQueryParams: {
+					routeQuery: {
 						type: Object,
+						value: defaultObject,
 						notify: true
 					},
 
-					/**
-					 */
 					routeHash: {
-						type: String,
-						notify: true
-					},
-
-					routeHashParams: {
 						type: Object,
+						value: defaultObject,
 						notify: true
 					},
 
-					_urlQueryString: {
-						type: String
-					},
-
-					_urlHash: {
-						type: String
-					},
-
-					_tempParamsObject: {
-						type: Object
-					},
-
-					_tempParamsString: {
-						type: String
+					hashBang: {
+						type: Boolean,
+						readOnly: true
 					}
 				},
 
 				observers: [
-					'_computeRoutePathAndParams(_urlHash)',
-					'_updateHash(routePath, routeQueryString, routeHash)'
+					'_routeChanged(routeQuery.*)',
+					'_routeChanged(routeHash.*)',
+					'_appHashChanged(_appHashString)',
```

Comment:
But this observer in the property declaration instead.

Reply:
@nomego This needs to be a complex observer that is notified of sub property updates.

Reply:
It isn't complex it observes _appHashString?

Reply:
For `_appHashString` it can be done. I meant for `routeQuery` and `routeHash`.